---
title: "10X Analysis Template: TCRs"
output: html_notebook
author: "Kirsten Diggins"
---

This template provides code to read, process, and plot TCR counts from 10X analysis. Code for TCR data reading and pre-processing was written by Hannah DeBurg, and airline plot code was written by Matt Dufort. The merging and plotting steps require a post-processed, clustered cds object from GEX analysis.

## Workflow:
1. Load and process TCR counts; get QC plots.
2. Merge TCR libs and run tcrGraph to get clonotypes.
3. Merge TCR data with GEX data for data exploration.
4. Plots to explore clonal expansion.


```{r load_libraries, message=FALSE, warning=FALSE}
library(dplyr)
library(ggthemes)
library(ggbeeswarm)
library(viridis)
library(stringr)
library(readxl)
library(kableExtra)
library(RColorBrewer)
library(plotly)
library(tidyr)
library(gtools)
library(apird) #API for the research database
#library(devtools)  #if needed to obtain github packages
# install_github('mjdufort/TCRtools') #if needed to get Matt Dufort's package
library(TCRtools) 
# library(annotables)
#devtools::install_github("stephenturner/annotables")
library(RNAseQC) #install_github('benaroyaresearch/RNAseQC')
library(data.table)
library(edgeR)
library(ggrepel)
library(ComplexHeatmap)
library(geneSetTools) # For barcode plots
library(egg) #For ggarrange
library(ggpubr) #Also for ggarrange
library(DGETools) #install_github('benaroyaresearch/DGETools')
library(inlmisc) #For colors
library(umap)
library(tcrGraph)
library(igraph)
library(forcats)
library(ggalluvial)
library(Seurat)
library(monocle3)
```

## Load and process TCR counts; get QC plots

```{r get_data_and_dirs}
## Set path to data files
# file_path = "my_path/"
file_path = "C:/Users/kdiggins/Box/P362-1 T1DAL 10X/200821-10X_T1DAL_Results/200821-10X_T1DAL_Results/"

## Load annotation file
# load(file_path,"project annotation.Rdata") ## project_anno
load("C:/Users/kdiggins/Box/P362-1 T1DAL 10X/P362-2 annotation.Rdata")
anno_data <- P362_anno ## rename 

## Load GEX data (cds object; post-processing with cluster and trajectory information)
# load(paste0(file_path,"cds_object.Rdata"))
load("C:/Users/kdiggins/Box/P362-1 T1DAL 10X/P362-1 T1DAL cds object - postQC no MAIT cells.Rdata")
cds_all <- cds_no_MAIT

## Filter annotation to just TCR libraries
# tcr_anno <- project_anno %>% dplyr::filter(libraryProtocolId=="next_gem_sc_5prime_10x_genomics_tcr_enriched") 
tcr_anno <- P362_anno %>% dplyr::filter(libraryProtocolId=="next_gem_sc_5prime_10x_genomics_tcr_enriched") ## library protocol ID

head(tcr_anno)
```
Read in individual TCR data files (anno, clonotypes, metrics) and merge into aggregated objects:

```{r read_10x_tcr_data}
## Get TCR library IDs
libs <- tcr_anno %>% dplyr::select(libid)
libs <- as.vector(libs$libid)

tcr_files <- file.path(file_path, "individualLibraryResults","tcr", paste0(libs, "_all_contig_annotations.csv"))
clonotype_files <- file.path(file_path, "individualLibraryResults","tcr", paste0(libs, "_clonotypes.csv"))
tcr_qc_files <- file.path(file_path, "individualLibraryResults","tcr", paste0(libs, "_metrics_summary.csv"))

unpack_file <- function(file_in){
  
  data <- read.csv(file_in)
  data$libid <- str_extract(file_in, "lib[0-9]+")
  
  return(data)
  
}

tcrs_list <- lapply(tcr_files, unpack_file)
tcrs <- bind_rows(tcrs_list)

clonotypes_list <- lapply(clonotype_files, unpack_file)
clonotypes <- bind_rows(clonotypes_list)

## Extract and bind all tcr metrics
tcr_metrics <- lapply(tcr_qc_files, unpack_file)

## Fix character conversion of estimated number of cells in files
make_char_numeric <- function(file_in){
  in_file <- file_in
  in_file$Mean.Read.Pairs.per.Cell <- as.numeric(str_remove(in_file$Mean.Read.Pairs.per.Cell,","))
  in_file$Mean.Read.Pairs.per.Cell <- as.numeric(str_remove(in_file$Mean.Read.Pairs.per.Cell,","))
  in_file$Estimated.Number.of.Cells <- as.numeric(str_remove(in_file$Estimated.Number.of.Cells,","))
  in_file$Number.of.Cells.With.Productive.V.J.Spanning.Pair <- as.numeric(str_remove(in_file$Number.of.Cells.With.Productive.V.J.Spanning.Pair,","))
  
  return(in_file)
}

tcr_metrics <- lapply(tcr_metrics,make_char_numeric)

tcr_metrics <- bind_rows(tcr_metrics)
head(tcr_metrics)
```

# Data quality metrics

```{r tcr_quality_cleaning}
make_pct_numeric <- function(pct_column){
  
  num_column <- pct_column %>%
    str_remove_all("%") %>%
    as.numeric() 
  num_column <- num_column/100
  
  return(num_column)
}

tcr_metrics <- tcr_metrics %>%
  dplyr::mutate(Reads.Mapped.to.Any.V.D.J.Gene = make_pct_numeric(Reads.Mapped.to.Any.V.D.J.Gene),
                Cells.With.Productive.TRA.Contig = make_pct_numeric(Cells.With.Productive.TRA.Contig),
                Cells.With.Productive.TRB.Contig = make_pct_numeric(Cells.With.Productive.TRB.Contig))

tcr_anno_metrics <- merge(tcr_anno,tcr_metrics,by="libid")
head(tcr_anno_metrics)
```

# Get TCR metrics plots

```{r}
ggplot(tcr_anno_metrics) +
  geom_col(aes(x = participantID,
                 y = Reads.Mapped.to.Any.V.D.J.Gene*100,
               fill = participantID))+
  labs(x = "",
       y = "Reads mapped to V/D/J genes (%)")
```

```{r}
ggplot(tcr_anno_metrics) +
  geom_col(aes(x = participantID,
                 y = Cells.With.Productive.TRA.Contig*100,
               fill = participantID))+
  labs(x = "",
       y = "Cells with a productive TRA contig (%)")
```

```{r}
ggplot(tcr_anno_metrics) +
  geom_col(aes(x = participantID,
                 y = Cells.With.Productive.TRB.Contig*100,
               fill = participantID))+
  labs(x = "",
       y = "Cells with a productive TRB contig (%)")
```

```{r save_TCR_data}
# save(tcrs,file=paste0(file_path,"per-cell TCR data - postQC.Rdata"))
```

## Merge TCR libs and run tcrGraph to get clonotypes

```{r load_saved_tcr_data, message=FALSE, warning=FALSE}
## Load per-cell, post-QC TCR data (generated by code above)
# load(paste0(file_path,"per-cell TCR data - postQC.Rdata"))

## Load project annotation if not loaded above
# load(paste0(file_path,"annotation.Rdata"))
# anno_data <- project_anno
```

Barcode IDs may be repeated across flow cells, so to be sure that each cell has a unique ID, first append cell barcode to donor ID and rename to "libid" (for use in tcrGraph). 

```{r get_TCR_libs}
## Relabel cells with barcode_DonorID to ensure unique cell IDs across samples.
anno_to_merge <- anno_data %>% dplyr::rename(Donor.ID = participantID) %>% select(libid, Donor.ID) 

all_libs_tcrs <-  tcrs %>% 
  dplyr::filter(is_cell=="True", productive=="True", cdr3 != "None") %>%  ## keep cells with productive TCR sequences
  merge(anno_to_merge) %>%
  dplyr::rename(Original.Lib.ID = libid) %>% 
  mutate(libid = paste(barcode,Donor.ID,sep="_"))
head(all_libs_tcrs)
```

# Run tcrGraph on combined samples to get global clonotype definitions. 

```{r run_tcrGraph}
## Run TCR graph to get global clonotype definitions across samples.
tcr_graph_output_all = tcrGraph::makeTcrGraph(all_libs_tcrs,link="cdr3_nt")

tcrGraph_clones_all <- tcrGraph::getClonesFromTcrGraph(tcr_graph_output_all, link = "cdr3_nt")

# length(unique(tcrGraph_clones_all$cloneId)) ## check number of clonotypes identified

tcrGraph::makeVisNetwork(tcr_graph_output_all)
```


## Merge TCR info with GEX data for data exploration

Add back tcrGraph clonotype ID to cds colData for plotting: reformat tcrGraph clone IDs and sequence information to merge with cds.

The barcode suffix (-1,-2, etc) designates the GEM well. The samples in this experiment were each run in a single GEM well, so in the raw data this should be just -1 across the board. However, in the GEX aggregation step, it seems to be iteratively increasing the suffix with each additional library. So, to merge TCR barcodes with aggregated GEX barcodes, itâ€™s necessary to remove the suffix.

```{r get_cloneIDs}
all_cloneID_data = data.frame(barcode_donorID = NULL, 
                             Donor.ID = NULL,
                             barcode = NULL, 
                             cloneID = NULL, 
                             clone_count = NULL,
                             vGenes = NULL,
                             jGenes = NULL,
                             cdr3_nt = NULL)

for(i in 1:length(unique(tcrGraph_clones_all$cloneId))){
  single_clone_data <- tcrGraph_clones_all[i,] 
  
  ## Extract IDs
  both_IDs = unlist(str_split(single_clone_data$libs, ", "))
  Donor.IDs =  sapply(strsplit(both_IDs, "_"), `[`, 2)
  barcodes_long = sapply(strsplit(both_IDs, "_"), `[`, 1)
  barcodes = sapply(strsplit(barcodes_long,"-"), `[`, 1)
  
  ## Build single clone data frame
  single_clone <- data.frame(barcode_donorID = paste(barcodes,Donor.IDs,sep="_"), 
                             Donor.ID = Donor.IDs,
                             barcode_original = barcodes_long,
                             barcode = barcodes,
                             cloneID = single_clone_data$cloneId, 
                             clone_count = single_clone_data$cloneCounts,
                             vGenes = single_clone_data$vGenes,
                             jGenes = single_clone_data$jGenes,
                             cdr3_nt = single_clone_data$cdr3_nt)
  
  all_cloneID_data <- bind_rows(all_cloneID_data,single_clone)
}
head(all_cloneID_data)
```

```{r reformat_cds_metadata}
## Reformat colData(cds_all) for merging with TCR data
colData(cds_all)$barcode_original <- row.names(colData(cds_all))
cds_metadata <- colData(cds_all) %>% as.data.frame() %>% select(Donor.ID,barcode_original)
cds_metadata$barcode <-  sapply(strsplit(cds_metadata$barcode_original,"-"), `[`, 1)
cds_metadata$barcode_donorID <- paste(cds_metadata$barcode,cds_metadata$Donor.ID,sep="_")
cds_metadata <- cds_metadata %>% select(-barcode) ## to avoid redundancy in merged colData
head(cds_metadata)
```
Create cds metadata object with clone information in same order, same dimensions as colData(cds). If you try to directly merge clone info with cds metadata and then add back the new colData(cds) data frame, this creates row name issues. I've found it easier to add columns to colData rather than replace the whole thing.

```{r merge_TCR_with_colData}
## Merge colData(cds_all) with TCR info by barcode and Donor ID; if there's a matching barcode and ID between GEX and TCR lib, it will merge, otherwise those fields will be NA
all_cloneID_data_to_merge <- all_cloneID_data %>% dplyr::select(-barcode_original,-Donor.ID)
cds_metadata_with_clones <- left_join(cds_metadata,all_cloneID_data_to_merge, by = "barcode_donorID") 

cds_metadata_with_clones
```
```{r}
## Check that barcode ID order is the same in the cds metadata object and in colData(cds_all)
all(cds_metadata_with_clones$barcode_original == row.names(colData(cds_all)))
## True = TCR and clonotype data is in the same order as the rows in colData(cds)
```

```{r add_new_colData}
## Since barcodes in metadata match exactly with colData(cds), no need to merge (removes row names); can just append desired columns to colData(cds)

colData(cds_all)$cloneID <- cds_metadata_with_clones$cloneID
colData(cds_all)$clone_count <- cds_metadata_with_clones$clone_count
colData(cds_all)$vGenes <- cds_metadata_with_clones$vGenes
colData(cds_all)$cdr3_nt <- cds_metadata_with_clones$cdr3_nt
colData(cds_all)$barcode_donorID <- cds_metadata$barcode_donorID

head(colData(cds_all))
```

## Plots for data exploration

# UMAP plot of cells colored by expansion

Add heat intensity overlay for TCR-expressing cells on UMAP based on clonal expansion.

```{r plot_expansion}
## Replace NA with 0 for clone counts
clone_counts <- colData(cds_all)$clone_count
clone_counts[is.na(clone_counts)] <- 0
colData(cds_all)$clone_count_numeric <- clone_counts
cds_metadata_ordered <- colData(cds_all) %>% 
  as.data.frame() %>% 
  arrange(clone_count_numeric,) ## reorder so cells with highest clone count are on top

cds_all_ordered <- cds_all[,cds_metadata_ordered$barcode_original]
plot_cells(cds_all_ordered,
           color_cells_by="clone_count",
           show_trajectory_graph = F, 
           cell_size=2) 
```
```{r}
plot_cells(cds_all_ordered,
           color_cells_by="clone_count",
           show_trajectory_graph = F, 
           cell_size=2) + facet_wrap(~ Response)
```

# Airline plots - connect cells on UMAP by shared clonotype.

This code was written by Matt Dufort. 

```{r}
## determine curves to link, based on clone_id_tcr_graph_clonal_expansion
colData(cds_all)$cluster <- colData(cds_all)$Cluster.Name
cds_pc_filtered_align_cdr = cds_all[,sample(colnames(cds_all),2000,replace=F)] ## Downsample large data sets
# make a copy of the annotation plus UMAP coordinates, for easier manipulation
data.tmp <-
  as.data.frame(colData(cds_pc_filtered_align_cdr)) %>%
  cbind(
    as.data.frame(reducedDims(cds_pc_filtered_align_cdr)$UMAP) %>%
      magrittr::set_colnames(c("V1", "V2"))) %>% 
  dplyr::rename("clone_id_tcr_graph_clonal_expansion" = "cloneID")

# create data frame to store links
curves.tmp <-
  data.frame(
    clone_id_tcr_graph_clonal_expansion = character(),
    x = numeric(),
    y = numeric(),
    xend = numeric(),
    yend = numeric())

# loop over each clone, and extract coordinates for cells from the same clone
for (clone_id.tmp in na.omit(unique(data.tmp$clone_id_tcr_graph_clonal_expansion))) {
  clone_id_curves.tmp <- curves.tmp[0,]
  data_for_curves.tmp <-
    data.tmp %>%
    dplyr::filter(clone_id_tcr_graph_clonal_expansion %in% clone_id.tmp)
  if (nrow(data_for_curves.tmp) > 1) {
    for (i in 1:(nrow(data_for_curves.tmp)-1)) {
      for (j in (i+1):nrow(data_for_curves.tmp)) {
        clone_id_curves.tmp <-
          rbind(
            clone_id_curves.tmp,
            list(
              clone_id_tcr_graph_clonal_expansion =
                data_for_curves.tmp$clone_id_tcr_graph_clonal_expansion[i],
              x = data_for_curves.tmp$V1[i],
              y = data_for_curves.tmp$V2[i],
              cluster1 = data_for_curves.tmp$cluster[i],
              xend = data_for_curves.tmp$V1[j],
              yend = data_for_curves.tmp$V2[j],
              cluster2 = data_for_curves.tmp$cluster[j]))
      }
    }
  }
  curves.tmp <-
    rbind(curves.tmp, clone_id_curves.tmp)
}

pal.cluster_renumbered = toupper(c('#a6cee3','#1f78b4','#b2df8a','#33a02c','#fb9a99','#e31a1c','#fdbf6f','#ff7f00','#cab2d6','#6a3d9a','#ffed6f','#b15928', "gray","black","blue","red")) 
# make the plot
plot_cells(
  cds_pc_filtered_align_cdr,
  color_cells_by = "cluster", cell_size=1, show_trajectory_graph = FALSE,
  group_label_size=8) +
  scale_color_manual(values=pal.cluster_renumbered) +
  geom_curve(
    data = curves.tmp,
    mapping = aes(x=x, y=y, xend=xend, yend=yend),
    size = 0.07,
    alpha=0.5) 
```
## Get heatmap of % shared cloneIDs between clusters.

```{r heatmap_shared_clones}
clusters <- unique(colData(cds_all)$Cluster.Name)
percent_shared <- matrix(nrow=length(clusters),ncol=length(clusters))
for(i in 1:length(clusters)){
  for(j in 1:length(clusters)){
    
    cluster_a <- dplyr::filter(as.data.frame(colData(cds_all)),Cluster.Name==i, !is.na(cloneID)) %>% select(cloneID) 
    cluster_b <- dplyr::filter(as.data.frame(colData(cds_all)),Cluster.Name==j, !is.na(cloneID)) %>% select(cloneID)
    
    shared <- intersect(cluster_a$cloneID,cluster_b$cloneID)
    total_clust_a_clonotypes <- length(unique(cluster_a$cloneID,na.rm=T))
    num_shared_clonotypes <- length(intersect(cluster_a$cloneID,cluster_b$cloneID))
    percent_shared[i,j] <- num_shared_clonotypes/total_clust_a_clonotypes*100
  }
}

colnames(percent_shared) <- clusters
row.names(percent_shared) <- clusters

Heatmap(percent_shared,cluster_rows = F, cluster_columns = F, column_names_rot = F, column_names_side = "top", row_names_side = "left")
```


